### ISSUE TITLE
Bem и Golang

### ISSUE THREAD

**Issue by Matveih1** [bem-site/bem-forum-content-ru#1288]
Добрый день. 
Подскажите: 
1. как подружить bem и golang?
2. как правильно передавать данные с сервера на клиента?

**Comment by tadatuta** [comment]
@Matveih1 нужно большой подробностей: что именно подразумевается под БЭМом в данном случае? Шаблонизатор?

**Comment by Matveih1** [comment]
Да, шаблонизатор. Т.е. как собрать страницу с помощью Gulp я разобрался, а вот как на нее передать данные с сервера (в данном случае на Go) не понимаю.

**Comment by zxqfox** [comment]
Можно поискать v8 биндинги для Go, но можно и использовать https://github.com/zxqfox/http-proxy-mitm, чтобы рендерить в HTML ответ от бекенда на Go.

Собранные шаблоны надо будет использовать как раз в проксе.

Есть вот такая демка для PHP на хероку, но бекендом может быть что угодно: https://github.com/zxqfox/php-fpm-bem-demo-heroku

**Comment by zxqfox** [comment]
По поводу второго: есть два варианта, либо это сырые данные и в дерево представления они превращаются в прокси, либо это сразу дерево и на него накладываются какие-нибудь bemhtml шаблоны. Возможны и смешанные варианты, когда часть или некоторые части представлены деревом.

На вкус и цвет, надо самому пробовать и разбираться что больше нравится.

**Comment by Matveih1** [comment]
спасибо, но как то не тривиально. Уверен, что есть способ проще. Хочется собрать HTML по методологии БЭМ, который сможет парсить GO и возвращать клиенту. Или клиент запрашивает у сервера данные для страницы и уже на стороне клиента их парсит. 
Наверно вопрос звучит по другому: как с помощью БЭМ создавать не статичные страницы, а динамичные. Действительно не важно что выступает в качестве бекенда. 
например как мне заполнить справочник данными из json файла.

**Comment by tadatuta** [comment]
Ключевой момент, который необходимо понимать: шаблонизатор BEMHTML работает на JS.
Так что если необходимо выполнять шаблоны на сервере, там потребуется либо node.js, либо еще какой-то вариант выполнения JavaScript. Если это требование устраивает, дальше все просто. В качестве примера см. https://github.com/bem/bem-express (там источником данных служит сервер на node.js, но принцип общий).

Если же подходит рендеринг на стороне браузера, то все еще проще:
1. В браузере собранные шаблоны подключаются как самый обычный JS-файл.
2. GO в ответ на AJAX-запрос отправляет на клиент данные в JSON.
3. На клиенте вызываем `BEMHTML.apply(myData)` и получаем HTML-строку, которую затем можно вставить в DOM.

**Comment by Matveih1** [comment]
Рендеринг на стороне браузера, устраивает:
1. Как попросить у GO JSON понятно
2. Как раскрутить его на клиенте - не понятно, есть готовые примеры?

**Comment by tadatuta** [comment]
Например, так: https://jsfiddle.net/tadatuta/h307ayjc/2/

**Comment by Matveih1** [comment]
Спасибо за ответ, но пока все еще не понятно. Теперь мне логика не понятна. Есть сервер который на запрос клиента должен отправить html ( собранный из bemjson.js - собранный заранее) + json который распарсится на клиенте и будет счастье или html который потом спросит у сервера json который опять же распасится на стороне клиента. Или как то по другому все происходит. 
Есть пример чата с помощью бэм? Или что то динамическое.

**Comment by tadatuta** [comment]
Между фиддлом, который я привел в предыдущем ответе и чатом будет крайне мало отличий в плане работы с шаблонами.
Общая идея такая:
1. Любым способом получить от сервера данные в JSON. В моем примере дергается API github, но это вполне может быть сокет в случае чата.
2. Передать эти данные в шаблонизатор и получить на выходе HTML-строку (`var html = BEMHTML.apply(data);`).
3. Вставить эту строку в DOM: (`bemDom.append(domElem, html);`)

**Comment by Matveih1** [comment]
Спасибо. Чувствую что уже близко, но не до конца понимаю. 1. Сервер вернет клиенту index.html который я собрал заранее с помощью gulp
2. Клиент попросит у сервера json с данными, который будет у себя парсить
Логика правильная или я уже на таком уровне не то говорю?

**Comment by Matveih1** [comment]
Спасибо. Чувствую что уже близко, но не до конца понимаю. 1. Сервер вернет клиенту index.html который я собрал заранее с помощью gulp
2. Клиент попросит у сервера json с данными, который будет у себя парсить
Логика правильная или я уже на таком уровне не то говорю?

**Comment by tadatuta** [comment]
Вполне норм
